### PR TITLE
feat: add stack backtrace

### PR DESCRIPTION
**Pull request by infdahai**
issue : #309

just move the [stack trace impl](https://github.com/chenshuo/muduo/blob/master/muduo/base/CurrentThread.cc#L22-L74) in muduo to bustub.

### CODE REVIEW

**Comment by skyzh** [comment]
Wow, this is really useful! Thanks a lot!

**Comment by infdahai** [comment]
> Wow, this is really useful! Thanks a lot!

I want to know whether gtests already provided some suitable backtrace in the repository?

**Comment by infdahai** [comment]
> > Wow, this is really useful! Thanks a lot!
> 
> I want to know whether gtests already provided some suitable backtrace in the repository?

I test the gtest output in error situations afterwards.

**Comment by infdahai** [comment]
for example,

**current assert behavior:**
```
[ RUN      ] LockManagerDeadlockDetectionTest.CycleTest1
~/workspace/bustub-private/test/concurrency/deadlock_detection_test.cpp:218: Failure
Expected equality of these values:
  1
  txn_id
    Which is: 4
[  FAILED  ] LockManagerDeadlockDetectionTest.CycleTest1 (50 ms)
```

**expected assert behavior:**
```
[ RUN      ] LockManagerDeadlockDetectionTest.CycleTest1
~/workspace/bustub-private/test/concurrency/deadlock_detection_test.cpp:218: Failure
Expected equality of these values:
  1
  txn_id
    Which is: 4

backtrace:
/workspace/.../a.cpp:100
/workspace/.../b.cpp:50

[  FAILED  ] LockManagerDeadlockDetectionTest.CycleTest1 (50 ms)
```

**Comment by infdahai** [comment]
in mutil-threads test, perhaps we don't get pretty backtrace in `assert failed` situation. it will print the same concurrent func multiple times and get the useless 'lib.so' call info.

File: src/include/execution/expressions/comparison_expression.h [hunk 1]
```diff
@@ -75,6 +76,7 @@ class ComparisonExpression : public AbstractExpression {
         return lhs.CompareGreaterThanEquals(rhs);
       default:
         BUSTUB_ASSERT(false, "Unsupported comparison type.");
+        return CmpBool::CmpFalse;
```
Comment (infdahai) [inline src/include/execution/expressions/comparison_expression.h:79]: I think it's better not to use assert() as the return value.

**Comment by skyzh** [comment]
after trying this on private repo, it seems that we cannot de-mangle symbols on macOS. Probably I'll prefer wait for another compiler version bump and use https://en.cppreference.com/w/cpp/utility/basic_stacktrace instead in C++ 23.

**Comment by infdahai** [comment]
ok, maybe this situation is related to https://github.com/NationalSecurityAgency/ghidra/issues/4652. so shall i close the pr?

**Comment by skyzh** [comment]
Let’s keep it and I’ll probably see if there’s any chance to integrate it.

**Pull request closed without merging**